homeassistant/components/recorder/__init__.py

```diff
@@ -0,0 +1,358 @@
+"""
+Support for recording details.
+
+Component that records all events and state changes. Allows other components
+to query this database.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/recorder/
+"""
+import logging
+import queue
+import threading
+import json
+import time
+from datetime import timedelta
+
+import voluptuous as vol
+
+import homeassistant.util.dt as dt_util
+from homeassistant.const import (EVENT_HOMEASSISTANT_START,
+                                 EVENT_HOMEASSISTANT_STOP, EVENT_STATE_CHANGED,
+                                 EVENT_TIME_CHANGED, MATCH_ALL)
+from homeassistant.core import Event, EventOrigin, State
+from homeassistant.helpers.event import track_point_in_utc_time
+
+DOMAIN = "recorder"
+
+REQUIREMENTS = ['sqlalchemy==1.0.13']
+
+DEFAULT_URL = "sqlite:///{hass_config_path}"
+DEFAULT_DB_FILE = "home-assistant_v2.db"
+
+CONF_DB_URL = "db_url"
+CONF_PURGE_DAYS = "purge_days"
+
+RETRIES = 3
+CONNECT_RETRY_WAIT = 10
+QUERY_RETRY_WAIT = 1
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_PURGE_DAYS): vol.All(vol.Coerce(int),
+                                               vol.Range(min=1)),
+        vol.Optional(CONF_DB_URL): vol.Url(''),
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+_INSTANCE = None
+_LOGGER = logging.getLogger(__name__)
+
+# These classes will be populated during setup()
+# pylint: disable=invalid-name
+Session = None
+
+
+def query_to_states(q):
+    """Given a query object, convert all results to an array of States."""
+    import sqlalchemy.exc
+    for _ in range(0, RETRIES):
+        try:
+            return [
+                row for row in
+                (row_to_state(row) for row in q)
+                if row is not None]
+        except sqlalchemy.exc.SQLAlchemyError as e:
+            log_error(e, retry_wait=QUERY_RETRY_WAIT, rollback=True)
+    return []
+
+
+def query_to_events(q):
+    """Query the database and return a list of states."""
+    import sqlalchemy.exc
+    for _ in range(0, RETRIES):
+        try:
+            return [
+                row for row in
+                (row_to_event(row) for row in q)
+                if row is not None]
+        except sqlalchemy.exc.SQLAlchemyError as e:
+            log_error(e, retry_wait=QUERY_RETRY_WAIT, rollback=True)
+    return []
+
+
+def row_to_state(row):
+    """Convert a database row to a state."""
+    try:
+        return State(
+            row.entity_id, row.state,
+            json.loads(row.attributes),
+            dt_util.UTC.localize(row.last_changed),
+            dt_util.UTC.localize(row.last_updated)
+        )
+    except ValueError:
+        # When json.loads fails
```

Comment:
If the field is truncated, munged, or otherwise damaged such as bitrot happens, yes, it can raise if the json is invalid, as the json comes from the stored string.